Do not realize all configurations in a project by default (#31682)
gradle/gradle: platforms/documentation/docs/src/docs/release/notes.md

Refer to [`ProjectLayout.getSettingsDirectory()`](org/gradle/api/file/ProjectLayout.html#getSettingsDirectory()) for additional details.

#### Configurations are initialized lazily

> Similar to [tasks](userguide/lazy_configuration.html), Configurations are now only realized when necessary.

Reviewer: ```suggestion
Similar to [tasks](userguide/lazy_configuration.html), [configurations](userguide/declaring_configurations.html) are now realized only when necessary.
```

> Starting with this release, applying the `base` plugin or any plugin that applies the `base` plugin (e.g., JVM plugins) will no longer realize all configurations declared using `register` or the incubating role-based factory methods.

Reviewer: ```suggestion
Starting with this release, applying the `base` plugin (or any plugin that applies it, such as JVM plugins) no longer realizes all configurations declared using `register` or the incubating role-based factory methods.
```

> In some builds, this will reduce configuration time and memory usage.

Reviewer: ```suggestion
This change can reduce configuration time and memory usage in some builds.
```

> Configurations should be declared using the `register` method instead of the `create` methods to have the largest impact.

Reviewer: ```suggestion
To leverage these improvements, declare configurations using the `register` method instead of the `create` method:
```

ADD RELEASE FEATURES BELOW
vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv -->

<a name="build-authoring"></a>
### Build authoring improvements

Gradle provides rich APIs for plugin authors and build engineers to develop custom build logic.

#### Configurations are initialized lazily

> Similar to Tasks, Configurations are now initialized lazily, only when necessary.

Reviewer: ```suggestion
Similar to [Tasks](userguide/lazy_configuration.html), Configurations are now initialized lazily when necessary.
```

Author: Reworded

> Now, when applying the `Base` plugin or any plugins derived from it such as the JVM plugins, configurations that are lazily declared using `register` or through the incubating role-based factory methods will no longer be eagerly realized unless required.

Reviewer: ```suggestion
Starting with this release, applying the `Base` plugin or any of its derived plugins (e.g., JVM plugins) ensures that configurations declared using `register` or the incubating role-based factory methods are no longer eagerly realized unless required.
```

Author: Reworded

> Configurations should be declared using the `register` method instead of the `create` method to take advantage of the lazy initialization.

Reviewer: Can we add a code snippet here? It will be more impactful for people that are glancing at the release notes.

Author: Added

Refer to [`ProjectLayout.getSettingsDirectory()`](org/gradle/api/file/ProjectLayout.html#getSettingsDirectory()) for additional details.

#### Configurations are initialized lazily

> Similar to [Tasks](userguide/lazy_configuration.html), Configurations are now initialized lazily when necessary.

Reviewer 2: ```suggestion
Similar to [tasks](userguide/lazy_configuration.html), Configurations are only realized when necessary. 
```

> Starting with this release, applying the `Base` plugin or any of its derived plugins (e.g., JVM plugins) will not cause configurations declared using `register` or using the incubating role-based factory methods to be eagerly realized unless required.

Reviewer 2: This should be `base` and possibly link to base_plugin user manual chapter. 

```suggestion
Starting with this release, applying the `base` plugin or any plugin that applies the `base` plugin (e.g., JVM plugins) will no longer realize all configurations declared using `register` or the incubating role-based factory methods. 

In some builds, this will reduce configuration time and memory usage. 
```

Reviewer 2: Do we have any way of quantifying the impact?

Author: I don't expect much atm. But, I would expect more when the JVM plugins switch over to registering role-locked configurations 9.0 (vs eagerly creating unlocked configurations)

> Configurations should be declared using the `register` method instead of the `create` method to take advantage of the lazy initialization.

Reviewer 2: ```suggestion
Configurations should be declared using the `register` method instead of the `create` methods to have the largest impact.
```

```kotlin
configurations {
    // Instead of using `create`
    create("myEagerConfiguration")
    

>     // Use `register` to lazily initialize the configuration

Reviewer 2: ```suggestion
    // Use `register` to avoid realizing the configuration
```
